### PR TITLE
305 Add sort option for the name query

### LINKED ISSUE #305: [API] - Sorting
MW parent job: `[OMSBEA] omSupply Back End - API`

### Description

Mechanism for sorting

### Scope

<ins>Includes</ins>
* For one query only (say names)

<ins>Excludes</ins>
* Multisort

### Tasks

- [ ] Break down tasks into digestible (by reviewer) units.
- [ ] MW job added

### PR DESCRIPTION
**Pull request by clemens-msupply**
Closes #305

### CODE REVIEW

File: src/database/repository/diesel/name_query.rs [hunk 1]
```diff
@@ -107,9 +118,28 @@ impl NameQueryRepository {
...
         }
 
-        let result = query
-            .order(name_table_dsl::id.asc())
-            .load::<NameAndNameStoreJoin>(&*connection)?;
+        if let Some(sort) = sort {
+            match sort.key {
+                NameQuerySortField::Name => {
+                    if sort.desc.unwrap_or(false) {
+                        query = query.order(name_table_dsl::name.desc());
```
Comment (andreievg) [inline src/database/repository/diesel/name_query.rs:125]: ohh man i wish there was a way to abstract this into a method, i tried a little bit, and gave up pretty quickly

**Review by andreievg — APPROVED** [review]
Looks good, approving, but i think we want to do a Option<Vec<NameSortOption>> as input in graphql

File: src/database/repository/diesel/name_query.rs [hunk 2]
```diff
@@ -54,6 +54,16 @@ pub struct NameQueryFilter {
     pub is_supplier: Option<bool>,
 }
 
+pub struct NameQuerySortOption {
```
[the remark this answers is not in the document]
Comment (wlthomson) [reply src/database/repository/diesel/name_query.rs:57]: I agree that this is probably fine just as `NameQuerySort`. Mostly just think the `Option` bit is unneeded, but also has some potential for confusion with the Rust enum.

File: src/database/repository/diesel/name_query.rs (hunk 1 above)
Comment (wlthomson) [reply src/database/repository/diesel/name_query.rs:125]: The only way I can think of is to somehow pass the DSL command as a closure to a method which then examined the value of `sort` to determine whether to append `.desc()` or `.asc()`. Doable in JS, but not surprised to hear it's not so easy in Rust 😆.

**Review by wlthomson — APPROVED** [review]

File: src/server/service/graphql/schema/types/name.rs [hunk 3]
```diff
@@ -62,6 +79,15 @@ impl NameList {
 
     async fn nodes(&self, ctx: &Context<'_>) -> Vec<NameQuery> {
         let repository = ctx.get_repository::<NameQueryRepository>();
-        repository.all(&self.pagination, &self.filter).unwrap()
+        repository
+            .all(
+                &self.pagination,
+                &self.filter,
+                &self.sort.as_ref().map(|opt| NameQuerySortOption {
```
[the remark this answers is not in the document]
Comment (clemens-msupply) [reply src/server/service/graphql/schema/types/name.rs:86]: Yes thought about it but its most likely only used once and I find it easier to read to have it explicit especially for such simple conversions...

File: src/database/repository/diesel/name_query.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (clemens-msupply) [reply src/database/repository/diesel/name_query.rs]: Renamed it. I also liked the idea to have everything in the API ending with Input, NameSortInput and just call it NameSort in the respository layer. Can rename in the next iteration.

File: src/server/service/graphql/schema/queries/mod.rs [hunk 4]
```diff
@@ -24,11 +24,13 @@ impl Queries {
         &self,
         _ctx: &Context<'_>,
         #[graphql(desc = "pagination (first and offset)")] page: Option<Pagination>,
-        #[graphql(desc = "filters")] filter: Option<NameFilter>,
+        #[graphql(desc = "filters option")] filter: Option<NameFilter>,
+        #[graphql(desc = "sort options")] sort: Option<NameSortOption>,
```
[the remark this answers is not in the document]
Comment (clemens-msupply) [reply src/server/service/graphql/schema/queries/mod.rs]: Done in 88430e3

**Pull request merged**